ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFArrayIntersect.java

```diff
@@ -43,19 +47,27 @@ public GenericUDFArrayIntersect() {
     ObjectInspector defaultOI = super.initialize(arguments);
     checkArgCategory(arguments, ARRAY2_IDX, ObjectInspector.Category.LIST, FUNC_NAME,
         org.apache.hadoop.hive.serde.serdeConstants.LIST_TYPE_NAME); //Array1 is already getting validated in Parent class
+    if (!ObjectInspectorUtils.compareTypes(arrayOI.getListElementObjectInspector(), ((ListObjectInspector) arguments[ARRAY2_IDX]).getListElementObjectInspector())) { // check if elements of arrays are comparable
+      throw new UDFArgumentTypeException(1, ERROR_NOT_COMPARABLE);
+    }
     return defaultOI;
   }
 
   @Override public Object evaluate(DeferredObject[] arguments) throws HiveException {
 
     Object array = arguments[ARRAY_IDX].get();
+    Object array2 = arguments[ARRAY2_IDX].get();
+    if (array == null) {
+      return null;
+    }
 
-    if (array == null || arrayOI.getListLength(array) <= 0) {
+    if (array2 == null) {
       return null;
     }
 
     List<?> retArray3 = ((ListObjectInspector) argumentOIs[ARRAY_IDX]).getList(array);
```

Comment:
If you already touch this code I would suggest to change the retArray3 to **resultArray** or just simple **result** I think it would be more expressive

Reply:
changed variable names to match their usage